### PR TITLE
Mining Startup Scripts

### PR DESCRIPTION
**Pull request by StriderDM**
## Description
Startup scripts which allow user to select if mining is enabled and if so whether to do normal mining or merged mining.

Linked to  PR #2427 

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I ran `cargo test` successfully before submitting my PR.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: common/src/configuration/bootstrap.rs [hunk 1]
```diff
@@ -113,6 +113,8 @@ pub struct ConfigBootstrap {
     /// This will clean out the orphans db at startup
     #[structopt(long, alias("clean_orphans_db"))]
     pub clean_orphans_db: bool,
+    #[structopt(long, alias("enable_mininf"))]
```
Comment (hansieodendaal) [inline common/src/configuration/bootstrap.rs]: There could be a spelling mistake here?
`enable_mining` vs `enable_mininf`

Comment (StriderDM) [reply common/src/configuration/bootstrap.rs]: Thanks, that definitely was.

**Review by hansieodendaal — COMMENTED** [review]
Hi there,

@mikethetike and @StriderDM,  I propose we merge #2427 before we merge this PR, as currently the base node and console wallet cannot share the same startup folder, and which may also require some changes to this PR.

1. I propose a similar approach be followed as in #2427 where the latest XMRig version will always be installed. In my mind, installation and configuration should actually come before running the code, and not be part of the runtime script. The required XMRig configuration could be captured in the json file as part of the installation, so a user does not need to input that information every time to be used as command-line arguments.

2. These scripts do not currently cater for initialization, which should be detected if already done, so the config files can be re-used.

3. I propose a bit more organized folder layout for runtime, maybe similar to the Windows runtime created by the Windows installer and starting scripts. If we assume the installation folder to be `~/.tari-testnet` for the illustration, then:

    ```
    ~/.tari-testnet             [license file, README, soft symbolic links to scripts]
    ~/.tari-testnet/runtime     [all executables, install scripts and runtime scripts]
    ~/.tari-testnet/wallet      [the base node wallet and console wallet databases]
    ~/.tari-testnet/config      [all config files, including the master XMRig config]
    ~/.tari-testnet/log         [all log files]
    ~/.tari-testnet/ridcully    [all other databases; blockchain, peer, dht]
    ```

4. For the runtime folder structure proposal above, a dedicated config toml will be needed, that should be part of the installation, as well as the use of the `--config "${config_path}/???.toml" --log_config "${config_path}/???.yml" --base-path .` command-line parameters for the base node and console wallet.

Regards.

File: common/src/configuration/bootstrap.rs [hunk 2]
```diff
@@ -128,6 +130,7 @@ impl Default for ConfigBootstrap {
             input_file: None,
             command: None,
             clean_orphans_db: false,
+            enable_mining: false,
```
Comment (stringhandler) [inline common/src/configuration/bootstrap.rs:133]: This doesn't seem to be hooked up to anything. Does this work?

You can also achieve this by setting the env var `TARI_BASE_NODE__RIDCULLY__ENABLE_MINING`

**Review by stringhandler — COMMENTED** [review]
Looks good in general, but I'm not sure that the enable-mining command arg is hooked up. I could be wrong.

In terms of the prompts I think there should be a bit more info.

Perhaps:
```
You can mine Tari blocks using SHA3 and/or Monero blocks using RandomX with merge mining. Find out more [link] 
Would you like to merge mine with Monero?
(Y/N)
Stagenet node: (default 184.283.xx.xx) // Enter selects default
Monero address: (default....)
XMRig is not installed. Would you like to download it? (Y/N)
Would you like to mine using the Tari Base Node with SHA3 (Y/N)
```

I think that the first time you run it, it should save the values to the config files, and then after should just start with the last saved values, unless called with (e.g.) `start_mining init`

File: common/src/configuration/bootstrap.rs (hunk 2 above)
Comment (StriderDM) [reply common/src/configuration/bootstrap.rs:133]: It is hooked up on line 153-157 of main.rs
```
    // enable-mining argument takes precedence over config setting
    if bootstrap.enable_mining {
        node_config.enable_mining = true;
    }
```

File: applications/tari_base_node/ubuntu/start_up.sh [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  gnome-terminal --working-directory=$PWD -- sh start_tor.sh
+}
+
+start_base_node() {
+  start_tor
+  echo "Starting Base Node"
+  if [ $sha3mining -eq 0 ]; then
+    gnome-terminal --working-directory=$PWD -- ./tari_base_node
+  else
+    gnome-terminal --working-directory=$PWD -- ./tari_base_node --enable_mining
```
Comment (hansieodendaal) [inline applications/tari_base_node/ubuntu/start_up.sh:16]:
```suggestion
  echo
  if [ ! -f "${config_path}/base_node_id.json" ] || [ ! -f "${config_path}/wallet_id.json" ]
  then
    if [ ! -f "${config_path}/base_node_id.json" ] 
    then
      echo Creating new "${config_path}/base_node_id.json";
    fi
    if [ ! -f "${config_path}/wallet_id.json" ] 
    then
      echo Creating new "${config_path}/wallet_id.json";
    fi
    ./tari_base_node --create_id --init --config "${config_path}/config.toml" --log_config "${config_path}/log4rs_base_node.yml" --base-path .
  else
    echo Using existing "${config_path}/base_node_id.json";
    echo Using existing "${config_path}/wallet_id.json";
  fi
  if [ ! -f "${config_path}/log4rs_base_node.yml" ]
  then
    echo Creating new "${config_path}/log4rs_base_node.yml";
    ./tari_base_node --init --config "${config_path}/config.toml" --log_config "${config_path}/log4rs_base_node.yml" --base-path .
  else
    echo Using existing "${config_path}/log4rs_base_node.yml";
  fi
  echo
  if [ $sha3mining -eq 0 ]; then
    gnome-terminal --working-directory=$PWD -- ./tari_base_node --config "${config_path}/config.toml" --log_config "${config_path}/log4rs_base_node.yml" --base-path .
  else
    gnome-terminal --working-directory=$PWD -- ./tari_base_node --config "${config_path}/config.toml" --log_config "${config_path}/log4rs_base_node.yml" --base-path . --enable_mining
```

File: applications/tari_base_node/ubuntu/start_up.sh [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  fi
+}
+
+start_console_wallet() {
+ echo "Starting Console Wallet"
+ gnome-terminal --working-directory=$PWD -- ./tari_console_wallet
```
Comment (hansieodendaal) [inline applications/tari_base_node/ubuntu/start_up.sh:22]:
```suggestion
  if [ ! -f "${config_path}/console_wallet_id.json" ]
  echo
  then
    echo Creating new "${config_path}/console_wallet_id.json";
    ./tari_console_wallet --create_id --init --config "${config_path}/config.toml" --log_config "${config_path}/log4rs_console_wallet.yml" --base-path .
  else
    echo Using existing "${config_path}/console_wallet_id.json";
  fi
  if [ ! -f "${config_path}/log4rs_console_wallet.yml" ]
  then
    echo Creating new "${config_path}/log4rs_console_wallet.yml";
    ./tari_console_wallet --init --config "${config_path}/config.toml" --log_config "${config_path}/log4rs_console_wallet.yml" --base-path .
  else
    echo Using existing "${config_path}/log4rs_console_wallet.yml";
  fi
  echo
  gnome-terminal --working-directory=$PWD -- ./tari_console_wallet --config "${config_path}/config.toml" --log_config "${config_path}/log4rs_console_wallet.yml" --base-path .
```

File: applications/tari_base_node/ubuntu/start_up.sh [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+start_merge_mining_proxy() {
+ echo "Starting Merge Mining Proxy"
+ gnome-terminal --working-directory=$PWD -- ./tari_merge_mining_proxy
```
Comment (hansieodendaal) [inline applications/tari_base_node/ubuntu/start_up.sh:27]:
```suggestion
  echo
  if [ ! -f "${config_path}/log4rs_merge_mining_proxy.yml" ]
  then
    echo Creating new "${config_path}/log4rs_merge_mining_proxy.yml";
    gnome-terminal --working-directory=$PWD -- ./tari_merge_mining_proxy --init --config "${config_path}/config.toml" --log_config "${config_path}/log4rs_merge_mining_proxy.yml" --base-path .
  else
    echo Using existing "${config_path}/log4rs_merge_mining_proxy.yml";
    gnome-terminal --working-directory=$PWD -- ./tari_merge_mining_proxy --config "${config_path}/config.toml" --log_config "${config_path}/log4rs_merge_mining_proxy.yml" --base-path .
  fi
  echo
```

File: applications/tari_base_node/ubuntu/start_up.sh [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+download_xmrig() {
+ echo "Downloading XMRig"
+ mkdir xmrig
+ curl -L -\# https://github.com/xmrig/xmrig/releases/download/v6.5.0/xmrig-6.5.0-focal-x64.tar.gz > xmrig.tar.gz && tar -xvf xmrig.tar.gz -C xmrig
```
Comment (hansieodendaal) [inline applications/tari_base_node/ubuntu/start_up.sh:33]:
```suggestion
  curl -L -\# https://github.com/xmrig/xmrig/releases/download/v6.6.2/xmrig-6.6.2-focal-x64.tar.gz > xmrig.tar.gz && tar -xvf xmrig.tar.gz -C xmrig
```

File: applications/tari_base_node/ubuntu/start_up.sh [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+ start_xmrig
```
Comment (hansieodendaal) [inline applications/tari_base_node/ubuntu/start_up.sh:34]:
```suggestion
```

File: applications/tari_base_node/ubuntu/start_up.sh [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    read yn
+    case $yn in
+        [Yy]* ) sha3mining=0; merged_mining; break;;
+        [Nn]* ) start_base_node; exit;;
+        * ) echo "Please answer yes or no.";;
+    esac
+ done
+}
+
+echo "Do you want to enable mining?"
```
Comment (hansieodendaal) [inline applications/tari_base_node/ubuntu/start_up.sh:73]:
```suggestion
config_path="./config"
echo "Do you want to enable mining?"
```

File: applications/tari_base_node/ubuntu/start_up.sh [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+ curl -L -\# https://github.com/xmrig/xmrig/releases/download/v6.5.0/xmrig-6.5.0-focal-x64.tar.gz > xmrig.tar.gz && tar -xvf xmrig.tar.gz -C xmrig
+ start_xmrig
+}
+
+start_xmrig() {
+ echo 'Please input Monero Wallet Address for XMRig'
+ read wallet_address
+ echo 'Please input Merge Mining Proxy Address for XMRig (Default: 127.0.0.1:7878)'
+ read proxy_address
+ echo "Starting XMRig"
```
Comment (hansieodendaal) [inline applications/tari_base_node/ubuntu/start_up.sh:42]:
```suggestion
  wallet_address_default="55LTR8KniP4LQGJSPtbYDacR7dz8RBFnsfAKMaMuwUNYX6aQbBcovzDPyrQF9KXF9tVU6Xk3K8no1BywnJX6GvZX8yJsXvt"
  proxy_address_default="127.0.0.1:7878"
  
  echo
  echo "Do you want to use the default Stagenet Monero Wallet Address for XMRig?"
  echo "   '$wallet_address_default'"
  while true; do
    read yn
    case $yn in
        [Yy]* ) wallet_address=$wallet_address_default; break;;
        [Nn]* ) echo 'Please input Monero Wallet Address for XMRig'; read wallet_address; exit;;
        * ) echo "Please answer yes or no.";;
    esac
  done

  echo
  echo "Do you want to use the default Merge Mining Proxy Address for XMRig?"
  echo "   '$proxy_address_default'"
  while true; do
    read yn
    case $yn in
        [Yy]* ) proxy_address=$proxy_address_default; break;;
        [Nn]* ) echo 'Please input Merge Mining Proxy Address for XMRig'; read proxy_address; exit;;
        * ) echo "Please answer yes or no.";;
    esac
  done

  echo
  echo "Starting XMRig"
  echo
```

**Review by hansieodendaal — COMMENTED** [review]
My code suggestions enhanced the logic a bit with checks and balances to make the experience more seamless for the user, and consistent with the default directory structure on Windows. I only tested this for Ubuntu; MacOS must still be done.

If I missed some of the places to comment, see this file:
[start_up.txt](https://github.com/tari-project/tari/files/5643533/start_up.txt)

**Comment by hansieodendaal** [comment]
I think this can be merged as is, then we can update the scripts with a follow-up PR.

**Review by stringhandler — APPROVED** [review]

**Pull request merged**